### PR TITLE
Async

### PR DESCRIPTION
**Pull request by mgramigna**
# Summary

This PR convertes the ELM execution portions of the library to `async`, allowing for future patient sources to implement `findRecords` in a way that uses asynchronous operations (e.g. reaching out to an HTTP server or a database). Since a `Retrieve` can theoretically happen just about anywhere in the tree of ELM, the root ELM Expression `exec*` functions is now `async`, so all subclasses also have an `async`implementation of `exec*`

# New Behavior

From an execution standpoint, the results should be exactly the same as before. The only new behavior is that calling `execute` on an instance of the `Executor` class must now be done with an `await` or a `.then` since the result is a `Promise`. 

E.g.

``` TypeScript
// define executor and psource
// ...

executor
  .exec(psource)
  .then(result => { /* ... */ })
  .catch(err => { /* ... */ })
```

# Code Changes

* Convert every `execute`, `exec`, and `execArgs` to async
* Convert all usages of the above in the code to use `await`
  * Used `Promise.all` in cases where dealing with arrays of expressions to execute
* Implemented custom sorting with an `async` comparator function
  * The `SortClause` class was previously using the built in `.sort` function and calling `exec` within the comparator function to get the -1/1/0 ordering for the two elements. Since `exec` had to be `await`ed, it became impossible to use the built in function without relying on a specific iteration order of the array in order to precompute all the orders, which is inadvisable. Implemented a custom `asyncMergeSort` function that does `O(nlogn)` sorting while supporting an `async` comparison function
  * Added unit tests for above new utility
* Converted all elm unit tests to use `await` pattern for getting results and Promise rejection pattern for testing errors
* Updated examples and documentation to use `async/await` or `.then`

# Testing Guidance

* All the usual checks should pass
* Output of examples should be identical to prior code's
* Link dependency into project that uses execution and replace with the suggested pattern from the documentation/examples and make sure nothing breaks

Pull requests into cql-execution require the following.
Submitter and reviewer should ✔ when done.
For items that are not-applicable, mark "N/A" and ✔.

[CDS Connect](https://cds.ahrq.gov/cdsconnect) and [Bonnie](https://github.com/projecttacoma/bonnie) are the main users of this repository.
It is strongly recommended to include a person from each of those projects as a reviewer.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] Code passes lint and prettier (hint: use `yarn run test:plus` to run tests, lint, and prettier)
- [x] All dependent libraries are appropriately updated or have a corresponding PR related to this change
- [x] `cql4browsers.js` built with `yarn run build:browserify` if source changed.

**Reviewer:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -247,20 +253,20 @@ defined in _test/elm/*/logical-test.js_.  For example:
...
 describe('And', () => {
-  this.beforeEach(() => {
+  this.beforeEach(function () {
     setup(this, data);
   });
 
-  it('should execute allTrue as true', () => {
-    this.allTrue.exec(this.ctx).should.be.true();
+  it('should execute allTrue as true', async function () {
+    (await this.allTrue.exec(this.ctx)).should.be.true();
```
Comment (natjoe4) [inline README.md:261]:
```suggestion
    should(await this.allTrue.exec(this.ctx))be.true();
```

File: README.md [hunk 2]
```diff
@@ -247,20 +253,20 @@ defined in _test/elm/*/logical-test.js_.  For example:
...
   });
 
-  it('should execute someTrue as false', () => {
-    this.someTrue.exec(this.ctx).should.be.false();
+  it('should execute someTrue as false', async function () {
+    (await this.someTrue.exec(this.ctx)).should.be.false();
```
Comment (natjoe4) [inline README.md:265]:
```suggestion
    should(await this.someTrue.exec(this.ctx))be.false();
```

File: README.md [hunk 3]
```diff
@@ -247,20 +253,20 @@ defined in _test/elm/*/logical-test.js_.  For example:
...
   });
 
-  it('should execute allFalse as false', () => {
-    this.allFalse.exec(this.ctx).should.be.false();
+  it('should execute allFalse as false', async function () {
+    (await this.allFalse.exec(this.ctx)).should.be.false();
```
Comment (natjoe4) [inline README.md:269]:
```suggestion
    should(await this.allFalse.exec(this.ctx)).be.false();
```

**Review by lmd59 — APPROVED** [review]

**Review by natjoe4 — APPROVED** [review]
Looks good to me! Ran some integration test through fqm-execution and did a visual review

**Review by natjoe4 — APPROVED** [review]
